components/raftstore/src/store/fsm/peer.rs

```diff
@@ -987,36 +988,95 @@ where
         syncer.report_for_self(self_report);
     }
 
-    // Call msg PrepareFlashback to stop the scheduling and RW tasks.
-    // Once called, it will wait for the channel's notification in FlashbackState to
-    // finish. We place a flag in the request, which is checked when the
-    // pre_propose_raft_command is called. Stopping tasks is done by applying
-    // the flashback-only command in this way, But for RW local reads which need
-    // to be considered, we let the leader lease to None to ensure that local reads
-    // are not executed.
+    // Call msg PrepareFlashback is to check if the persistent state is present in
+    // RegionLocalState. Once called, the persistent state and memory state will
+    // first be checked and if neither is found, then a Raft Admin Command will be
+    // proposed to persist the state in the RegionLocalState to lock the region.
+    // Once this command is submitted, a callback will be invoked to call msg
+    // PrepareFlashback again to wait for the new change to be applied completely.
+    // So we can perform the flashback safely later.
     fn on_prepare_flashback(&mut self, ch: Sender<bool>) {
+        let region_id = self.region().get_id();
         info!(
             "prepare flashback";
-            "region_id" => self.region().get_id(),
+            "region_id" => region_id,
             "peer_id" => self.fsm.peer.peer_id(),
         );
-        if self.fsm.peer.flashback_state.is_some() {
-            ch.send(false).unwrap();
-            return;
+        // check persist state
+        let state_key = keys::region_state_key(region_id);
+        if let Some(target_state) = self
+            .ctx
+            .engines
+            .kv
+            .get_msg_cf::<RegionLocalState>(CF_RAFT, &state_key)
+            .unwrap()
+        {
+            info!(
+                "Prepare Flashback, already has a flashback local state:{:?} for {:?} ",
+                target_state.get_flashback_state(),
+                self.region(),
+            );
+            if target_state.get_flashback_state() == FlashbackState::Prepare {
```

Comment:
We could make `flashback_state` optional, so it's more clear to tell that the state is `None` rather than `Prepare`.